### ISSUE TITLE
Upload to storage with PUT – CORS issue

### ISSUE THREAD

**Issue by knyttl** [googleapis/java-storage#619]
I have a following CORS set for my bucket:

```
gsutil cors get gs://my-bucket
[
   {
   "maxAgeSeconds": 3600, 
   "method": ["POST", "GET", "PUT", "OPTIONS"], 
   "origin": ["*"], 
   "responseHeader": ["Authorization", "Content-Length", "Content-Range", "Content-Type", "Date", "Origin", "Server", "Transfer-Encoding", "Access-Control-Allow-Origin"]
  }
]
```

I generated the uploadUrl as described here:

https://cloud.google.com/storage/docs/access-control/signing-urls-with-helpers#storage-signed-url-object-java

When I try to upload from frontend, I can see the `OPTIONS` preflight request gets exactly the same headers as described above.

However, the `PUT` request does not contain any `Access-Allow-*` headers and therefore fails with:

```
Access to XMLHttpRequest at 'https://storage.googleapis.com/***/1605283190251' from origin 'https://my.origin' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

The documentation does not mention anything special to setup these – can you please describe what I am doing wrong?

**Comment by suraj-qlogic** [comment]
@knyttl,Thanks for filling this issue.Things you missed like if you want to upload a file through signed URL than you have to pass 
```Content-Type,Access-Control-Allow-Origin``` as a headers argument in your request. 
```javascript
headers: {
       'Access-Control-Allow-Origin': '*',
       'Content-Type': 'application/octet-stream'
}
```

Would you try below step and Please let us know if this helps you?

**CORS config for my bucket is looks like below::**
gsutil cors get gs://my-test-bucket

**Output::**
```json
[{
    "maxAgeSeconds": 3600, 
    "method": ["POST", "GET", "PUT", "OPTIONS"], 
    "origin": ["*"], 
    "responseHeader": ["Authorization", "Content-Length", "Content-Range", "Content-Type", "Date", "Origin", "Server", "Transfer-Encoding", "Access-Control-Allow-Origin"]
}]
```

**First generate signed URLs by using below java code:**

```java
package com.test.storage;

import com.google.cloud.storage.BlobId;
import com.google.cloud.storage.BlobInfo;
import com.google.cloud.storage.HttpMethod;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageException;
import com.google.cloud.storage.StorageOptions;

import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class GenerateV4GetObjectSignedUrl {
  private static final String projectId = "your-project-id";
  private static final String bucketName = "your-bucket-name";
  private static final String objectName = "your-object-name";

  public static void main(String[] args) {
    //
    generateV4GetObjectSignedUrl(projectId, bucketName, objectName);
  }

  public static void generateV4GetObjectSignedUrl(
      String projectId, String bucketName, String objectName) throws StorageException {

    Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();

    // Define resource
    BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucketName, objectName)).build();

    // Generate Signed URL
    Map<String, String> extensionHeaders = new HashMap<>();
    extensionHeaders.put("Content-Type", "application/octet-stream");

    URL url =
        storage.signUrl(
            blobInfo,
            5,
            TimeUnit.HOURS,
            Storage.SignUrlOption.httpMethod(HttpMethod.PUT),
            Storage.SignUrlOption.withExtHeaders(extensionHeaders),
            Storage.SignUrlOption.withV4Signature());

    System.out.println("Generated PUT signed URL:");
    System.out.println("URL::" + url);
    System.out.println("You can use this URL with any user agent, for example:");
    System.out.println(
        "curl -X PUT -H 'Content-Type: application/octet-stream' --upload-file my-file '"
            + url
            + "'");
  }
}

```

**Output::**
Generated PUT signed URL:
URL::generate Url
You can use this URL with any user agent, for example:
curl -X PUT -H 'Content-Type: application/octet-stream' --upload-file my-file 'genrated url'

**Now using front end upload file index.html**
```html
<form method="POST" enctype="multipart/form-data" id="fileUploadForm">
    <input type="text" name="extraField"/><br/><br/>
    <input type="file" name="files"/><br/><br/>
    <input type="submit" value="Submit" id="btnSubmit"/>
</form>
```
**jquery code::**
```javascript
$(document).ready(function () {
    $("#btnSubmit").click(function (event) {
        //stop submit the form, we will post it manually.
        event.preventDefault();
        fire_ajax_submit();
    });

});

function fire_ajax_submit() {
    // Get form
    var form = $('#fileUploadForm')[0];
    var data = new FormData(form);
    $("#btnSubmit").prop("disabled", true);
    $.ajax({
        type: "PUT",
        url:"generate-signed-url-goes-here",
        data: data,
        processData: false, //prevent jQuery from automatically transforming the data into a query string
        //Add this line to resolve cross origin error
        headers: {
               'Access-Control-Allow-Origin': '*',
               'Content-Type': 'application/octet-stream'
            },
        cache: false,
        timeout: 600000,
        success: function (data) {
            $("#result").text(data);
            console.log("SUCCESS : ", data);
            $("#btnSubmit").prop("disabled", false);
        },
        error: function (e) {
            $("#result").text(e.responseText);
            console.log("ERROR : ", e);
            $("#btnSubmit").prop("disabled", false);
        }
    });
 }   
```

**Comment by knyttl** [comment]
I found out the initial cause:

I used `extensionHeaders.put("Content-Type", "application/octet-stream");`, but then I uploaded with `image/png`. The server responded with `400 Bad Request`, but!!! it didn't send the CORS headers, so browser couldn't see the payload and just reported it as CORS disallowed. This mislead me to think that this is a CORS issue while this was the a wrong content  type.

The whole issue wouldn't happen if the CORS headers would also be added to the 400 response.

**Comment by suraj-qlogic** [comment]
@knyttl ,Thanks for reply,Would you please share your image upload code.

**Comment by knyttl** [comment]
The code is just taken from the official example. The problem arises if the content type is different on BE and FE and your server returns `400` which will be hidden in the browser as I state above.

```kotlin
    fun uploadUrl(objectName: String, contentType: String): String = storage
        .signUrl(
            BlobInfo.newBuilder(BlobId.of(bucketName, objectName)).build(),
            15,
            TimeUnit.MINUTES,
            Storage.SignUrlOption.httpMethod(HttpMethod.PUT),
            Storage.SignUrlOption.withExtHeaders(
                mapOf(
                    "Content-Type" to contentType,
                )
            ),
            Storage.SignUrlOption.withV4Signature()
        )
        .toString()
```

**Comment by suraj-qlogic** [comment]
@knyttl ,Thanks for this but would you share FE side code?

**Comment by frankyn** [comment]
Hi @knyttl,

Is this still an open issue for you?

**Comment by knyttl** [comment]
> Hi @knyttl,
> 
> Is this still an open issue for you?

As I state above:

I found out the initial cause:

I used extensionHeaders.put("Content-Type", "application/octet-stream");, but then I uploaded with image/png. The server responded with 400 Bad Request, but!!! it didn't send the CORS headers, so browser couldn't see the payload and just reported it as CORS disallowed. This mislead me to think that this is a CORS issue while this was the a wrong content type.

**The whole issue wouldn't happen if the CORS headers would also be added to the 400 response.**
